_posts/2016-06-29-concurrency-in-go.md

```diff
@@ -0,0 +1,69 @@
+
+The behaviour for ``GetAllTheThings`` is to fetch data from all services defined and combined them into a list. Let’s start with a naive approach.
+
+<script src="https://gist.github.com/gronnbeck/6caa0bac97e217e9542d41ab35398da4.js"></script>
+
+Above we see an example implementation of the naive approach. In other words we query each service sequentially. That means that the call to the Github service has to wait for the Facebook service. And the Twitter service needs to wait on both Github and Facebook. Since each of these services are not dependent on each other. We can improve this by performing the requests concurrently. Enter channels and goroutines.
+
+<script src="https://gist.github.com/gronnbeck/4feddc7018cd917aceea0e2d471bc978.js"></script>
+
+We’ve now modified the naive approach using channels and goroutines. We see that each call is being issued inside a goroutine. And that the results are being collected in the for-loop ant the end. The code can be read sequentially and therefore easy to reason about. It’s also explicitly concurrent since we explicitly issue several goroutines. The only caveat is that the results may not be returned in the same order as the routines were issued.
+
+(PS: I’ve ignored handling errors in the concurrent examples. Don’t do this at home. It’s just for pure readability).
+
+Notice that we can still are able to use the naive approach for fetching a resource: ``naive.Get(path string)``. And that the signature of the function is exactly the same as before. That is powerful! But does it actually run faster?
+
+In ``main.go`` we put everything together and do measure execution time to see if its actually faster.
+
+<script src="https://gist.github.com/gronnbeck/f40f95750c5ace8337afe03b1664c275.js"></script>
+
+<script src="https://gist.github.com/gronnbeck/6c58e8cec31565c5f70f96577b93b327.js"></script>
+
+The conclusion is yes, it runs faster. Actually, it runs an order of magnitude faster. If you want to run these experiments your self or just curious about the implementation. The full example project can be found [here](https://github.com/gronnbeck/concurrencypatterns).
+
+## Closing notes
+
+We have shown the it’s easy to utilise concurrency in Go using channels and goroutines. However, this post has simplified a lot and the caveats you may encounter using channels and goroutines are not fully dressed here. So use channels and goroutines with caution. They can cause a lot of headache if over used. The general advice is to always start by building something naive before optimising.
+
+I hope you have enjoyed reading this post. If I’ve done something completely absurd and stupid. Please tell me so, in the comment below or on twitter ([@gronnbeck](https://twitter.com/gronnbeck)). I’m still learning and having fun with Go. And I’m eager to learn from you as well.
```

Comment:
Think it is okay to tone down the `I’ve done something completely absurd and stupid.`